packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // A GestureDetector allows the detection of panning and zooming gestures on
+    // the child.
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        _constraints = constraints;
+        return Listener(
+          onPointerSignal: _receivedPointerSignal,
+          child: GestureDetector(
+            behavior: HitTestBehavior.opaque, // Necessary when translating off screen
+            onScaleEnd: _onScaleEnd,
+            onScaleStart: _onScaleStart,
+            onScaleUpdate: _onScaleUpdate,
+
+            child: Transform(
+              transform: widget.transformationController.value,
+              child: KeyedSubtree(
+                key: _childKey,
+                child: widget.child,
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// A thin wrapper on ValueNotifier whose value is a [Matrix4] representing a
+/// transformation.
+///
+/// The [value] defaults to the identity matrix, which corresponds to no
+/// transformation.
+///
+/// See also:
+///  * [InteractiveViewer.transformationController]
+class TransformationController extends ValueNotifier<Matrix4> {
+  /// Create an instance of [TransformationController].
+  ///
+  /// The [value] defaults to the identity matrix, which corresponds to no
+  /// transformation.
+  TransformationController([Matrix4 value]) : super(value ?? Matrix4.identity());
+
+  /// Return the scene point at the given viewport point.
+  ///
+  /// A viewport point is relative to the parent while a scene point is relative
+  /// to the child, regardless of transformation. Calling toScene with a
+  /// viewport point essentially returns the scene coordinate that lies
+  /// underneath the viewport point given the transform.
+  ///
+  /// The viewport transforms as the inverse of the child (i.e. moving the child
+  /// left is equivalent to moving the viewport right).
+  ///
+  /// This method is often useful when determining where an event on the parent
+  /// occurs on the child. This example shows how to determine where a tap on
+  /// the parent occurred on the child.
+  ///
+  /// ```dart
+  /// @override
+  /// void build(BuildContext context) {
+  ///   return GestureDetector(
+  ///     onTapUp: (TapUpDetails details) {
+  ///       _childWasTappedAt = _transformationController.toScene(
+  ///         details.localPosition,
+  ///       );
+  ///     },
+  ///     child: InteractiveViewer(
+  ///       transformationController: _transformationController,
+  ///       child: child,
+  ///     ),
+  ///   );
+  /// }
+  /// ```
+  Offset toScene(Offset viewportPoint) {
+    // On viewportPoint, perform the inverse transformation of the scene to get
+    // where the point would be in the scene before the transformation.
+    final Matrix4 inverseMatrix = Matrix4.inverted(value);
```

Comment:
Not sure if there's a situation where value can't be inverted or if we should fail gracefully in that case. Probably OK for now.

Reply:
That's an interesting thought.  I've seen these matrix operations start spitting out NaNs when there are infinite values in the matrix for example, which was the reason I prevented partially infinite boundaryMargins.  I'll keep an eye out for other problems coming from stuff like this.